Docs: GitOps reference

noahtalerman opened the pull request:
Noah: Big PR alert! For folks reviewing, if it's helpful for me to walk you through the changes on a call, please feel free to schedule some time w/ me.

---

Docs for Fleet's best practice GitOps: #13643 (also #17043)

Changes:
- New "GitOps" page replace "Configuration files" page. GitOps includes a reference of all top-level keys and their options
- Add "GitOps" page under "Using Fleet > The basics" for now. Might want to pull it to the top level in next iteration.
- Update broken links and add new redirects for the UI
- New "GitOps" page links to new "fleetctl apply" contributors page. "fleetctl apply" page explains that `fleetctl apply` is used for imports and backwards compatibility GitOps.
- "Organization settings" section in new "fleetctl-apply" page points to `PATCH /config` API endpoint as reference to cut down on doc content. 

Linking to this closed PR for commit history: #17890

rachaelshaw commented on docs/Contributing/Configuration-for-contributors.md:
```diff
@@ -1,367 +1,11 @@
 - [Environment variables](#environment-variables)
 
-This document includes configuration files and settings that are helpful when developing or contributing to Fleet.
+This document includes Fleet server confiurguration that are helpful when developing or contributing to Fleet.
```
```suggestion
This document includes Fleet server configuration settings that are helpful when developing or contributing to Fleet.
```

getvictor commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,550 @@
+agent_options:
+controls:
+org_settings: # Used in default.yml
+team_settings: # Used in teams/team-name.yml
```
`name` is also required for `teams/team-name.yml`

getvictor commented on docs/Using Fleet/GitOps.md:
````diff
@@ -0,0 +1,550 @@
+```yaml
+policies:
+  - path: ../lib/policies-name.policies.yml
+# path is relative to teams/team-name.yml 
````
path is relative to default.yml or teams/team-name.yml

getvictor commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,550 @@
+
+#### secrets
+
+The `secrets` section defines the valid secrets that hosts can use to enroll to Fleet. Supply one of these secrets when generating the fleetd agent you'll use to enroll hosts. Learn more [here](../Using%20Fleet/enroll-hosts.md).
```
`secrets` is an array/list of secrets, like:
```
  secrets:
    - secret: SampleSecret123
    - secret: ABC
```

Also, leaving `secrets` empty will mean the team will have no enroll secrets.

noahtalerman replied:
Thanks @getvictor! Addressed your feedback in my latest commit.

getvictor reviewed (commented):
Left a couple comments.

noahtalerman commented:
Hey @rachaelshaw, any way I can help with review on this PR? The merge conflict frequency is starting ti increase 📈

rachaelshaw commented:
@noahtalerman in the process of reviewing the new pages now & leaving a few comments. We should be able to get this merged today, sorry for the delay!

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+#### Options
+
+For possible options, see the parameters for the [Add policy API endpoint](../REST%20API/rest-api.md#add-policices).
```
```suggestion
For possible options, see the parameters for the [Add policy API endpoint](https://fleetdm.com/docs/rest-api/rest-api#add-policy).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+#### Options
+
+For possible options, see the parameters for the [Create query API endpoint](../REST%20API/rest-api.md#create-query).
```
```suggestion
For possible options, see the parameters for the [Create query API endpoint](https://fleetdm.com/docs/rest-api/rest-api#create-query).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+The `controls` section allows you to configure scripts and device management (MDM) features in Fleet.
+
+- `scripts` is a list of paths to macOS, Windows, or Linux scripts.
+- `windows_enabled_and_configured` specifies whether or not to turn on Windows MDM features (default: `false`). Can only be configure for all teams (`default.yml`).
```
```suggestion
- `windows_enabled_and_configured` specifies whether or not to turn on Windows MDM features (default: `false`). Can only be configured for all teams (`default.yml`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+- `scripts` is a list of paths to macOS, Windows, or Linux scripts.
+- `windows_enabled_and_configured` specifies whether or not to turn on Windows MDM features (default: `false`). Can only be configure for all teams (`default.yml`).
+- `enable_disk_encryption` specifies whether or not to enforce disk encryption on macOS and Widnows hosts (default: `false`).
```
```suggestion
- `enable_disk_encryption` specifies whether or not to enforce disk encryption on macOS and Windows hosts (default: `false`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+#### macos_updates
+
+- `deadline` specifies the deadline in the form of `YYYY-MM-DD`. The exact deadline time is at 04:00:00 (UTC-8) (default: empty).
```
Would clarify what "empty" means (`""`, `null`, etc.)
```suggestion
- `deadline` specifies the deadline in the form of `YYYY-MM-DD`. The exact deadline time is at 04:00:00 (UTC-8) (default: `""`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+#### macos_updates
+
+- `deadline` specifies the deadline in the form of `YYYY-MM-DD`. The exact deadline time is at 04:00:00 (UTC-8) (default: empty).
+- `minimum_version` specifies the minimum required macOS version (default: empty).
```
```suggestion
- `minimum_version` specifies the minimum required macOS version (default: `""`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+#### windows_updates
+
+- `deadline_days` (default: empty)
+- `grace_period_days` (default: empty)
```
Would also clarify the exact default values instead of "empty" (`0` or `null` I'd assume?)
```suggestion
- `deadline_days` (default: empty)
- `grace_period_days` (default: empty)
```

noahtalerman replied:
I think I confirmed that the default value is `null` by looking at the API response in dev console:

![Screenshot 2024-07-01 at 9 41 41 AM](https://github.com/fleetdm/fleet/assets/47070608/7033f77a-e983-4d94-9d81-5acb31cdc3a6)

![Screenshot 2024-07-01 at 9 41 33 AM](https://github.com/fleetdm/fleet/assets/47070608/1df0d7d9-500a-47a6-85a6-d20f07f456ea)

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+#### macos_setup
+
+The `macos_setup` section lets you control the [end user migration workflow](../Using%20Fleet/MDM-migration-guide.md#end-user-workflow) for macOS hosts that automatically enrolled to your old MDM solution.
```
```suggestion
The `macos_setup` section lets you control the [end user migration workflow](https://fleetdm.com/docs/using-fleet/mdm-migration-guide#end-user-workflow) for macOS hosts that automatically enrolled to your old MDM solution.
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+The `macos_setup` section lets you control the [end user migration workflow](../Using%20Fleet/MDM-migration-guide.md#end-user-workflow) for macOS hosts that automatically enrolled to your old MDM solution.
+
+- `bootstrap_package` is the URL to a bootstap package. Fleet will download the bootstrap package (default: empty).
```
```suggestion
- `bootstrap_package` is the URL to a bootstap package. Fleet will download the bootstrap package (default: `""`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+#### macos_migration
+
+- `enable` specifies whether or not to enable end user migration workflow (default: `false`)
+- `mode` specifies whether the end user initiates migration (`voluntary`) or they're nudged every 15-20 minutes to migrate (`forced`) (default: empty).
```
```suggestion
- `mode` specifies whether the end user initiates migration (`voluntary`) or they're nudged every 15-20 minutes to migrate (`forced`) (default: `""`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+Direct end users to a custom URL when they select **Transparency** in the Fleet Desktop dropdown (default: [https://fleetdm.com/transparency](https://fleetdm.com/transparency)).
+
+Can only be configure for all teams (`org_settings`).
```
```suggestion
Can only be configured for all teams (`org_settings`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+#### org_info
+
+- `name` is the name of your organization (default: empty)
```
```suggestion
- `name` is the name of your organization (default: `""`)
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+- `name` is the name of your organization (default: empty)
+- `logo_url` is a public URL of the logo for your organization (default: Fleet logo).
+- `org_logo_url_light_background` is a public URL of the logo for your organization that can be used with light backgrounds (default: Fleet logo).
+- `contact_url` is a URL that appears in error messages presented to end users (default: https://fleetdm.com/company/contact)
```
```suggestion
- `contact_url` is a URL that appears in error messages presented to end users (default: `"https://fleetdm.com/company/contact"`)
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+- `org_logo_url_light_background` is a public URL of the logo for your organization that can be used with light backgrounds (default: Fleet logo).
+- `contact_url` is a URL that appears in error messages presented to end users (default: https://fleetdm.com/company/contact)
+
+Can only be configure for all teams (`org_settings`).
```
```suggestion
Can only be configured for all teams (`org_settings`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+#### secrets
+
+The `secrets` section defines the valid secrets that hosts can use to enroll to Fleet. Supply one of these secrets when generating the fleetd agent you'll use to enroll hosts. Learn more [here](../Using%20Fleet/enroll-hosts.md).
```
```suggestion
The `secrets` section defines the valid secrets that hosts can use to enroll to Fleet. Supply one of these secrets when generating the fleetd agent you'll use to enroll hosts. Learn more [here](https://fleetdm.com/docs/using-fleet/enroll-hosts).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+#### server_settings
+
+- `enable_analytics` specifies whether or not to enable Fleet's [usage statistics](../Using%20Fleet/Usage-statistics.md) (default: `true`).
```
```suggestion
- `enable_analytics` specifies whether or not to enable Fleet's [usage statistics](https://fleetdm.com/docs/using-fleet/usage-statistics) (default: `true`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+- `scripts_disabled` blocks access to run scripts. Scripts may still be added in the UI and CLI (defaul: `false`).
+- `server_url` is the base URL of the Fleet instance (default: provided during Fleet setup)
+
+Can only be configure for all teams (`org_settings`).
```
```suggestion
Can only be configured for all teams (`org_settings`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
````diff
@@ -0,0 +1,548 @@
+    server_url: https://instance.fleet.com
+  ```
+
+TODO is SMTP down from new fleetctl-apply.md doc. Then gut the fleetctl-apply doc and point to this one.
````
Stray TODO. I think we can delete it?
```suggestion
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+The `sso_settings` section lets you define single sign-on (SSO) settings. Learn more about SSO in Fleet [here](https://fleetdm.com/docs/deploying/configuration#configuring-single-sign-on-sso).
+
+- `enable_sso` (default: `false`)
+- `idp_name` is the human-friendly name for the identity provider that will provide single sign-on authentication (default: empty).
```
```suggestion
- `idp_name` is the human-friendly name for the identity provider that will provide single sign-on authentication (default: `""`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+- `enable_sso` (default: `false`)
+- `idp_name` is the human-friendly name for the identity provider that will provide single sign-on authentication (default: empty).
+- `entity_id` is the entity ID is a Uniform Resource Identifier (URI) that you use to identify Fleet when configuring the identity provider. It must exactly match the Entity ID field used in identity provider configuration (default: empty `false`).
```
```suggestion
- `entity_id` is the entity ID: a Uniform Resource Identifier (URI) that you use to identify Fleet when configuring the identity provider. It must exactly match the Entity ID field used in identity provider configuration (default: `""`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+- `enable_sso` (default: `false`)
+- `idp_name` is the human-friendly name for the identity provider that will provide single sign-on authentication (default: empty).
+- `entity_id` is the entity ID is a Uniform Resource Identifier (URI) that you use to identify Fleet when configuring the identity provider. It must exactly match the Entity ID field used in identity provider configuration (default: empty `false`).
+- `metadata` is the metadata (in XML format) provided by the identity provider. (default: empty)
```
```suggestion
- `metadata` is the metadata (in XML format) provided by the identity provider. (default: `""`)
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+- `idp_name` is the human-friendly name for the identity provider that will provide single sign-on authentication (default: empty).
+- `entity_id` is the entity ID is a Uniform Resource Identifier (URI) that you use to identify Fleet when configuring the identity provider. It must exactly match the Entity ID field used in identity provider configuration (default: empty `false`).
+- `metadata` is the metadata (in XML format) provided by the identity provider. (default: empty)
+- `metadata_url` is the URL that references the identity provider metadata. Only one of  `metadata` or `metadata_url` is required (default: empty).
```
```suggestion
- `metadata_url` is the URL that references the identity provider metadata. Only one of  `metadata` or `metadata_url` is required (default: `""`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+- `enable_jit_provisioning` specified whether or not to allow single sign-on login initiated by identity provider (default: `false`). 
+- `enable_sso_idp_login` specifies whether or not to enables [just-in-time user provisioning](https://fleetdm.com/docs/deploy/single-sign-on-sso#just-in-time-jit-user-provisioning) (default: `false`).
+
+Can only be configure for all teams (`org_settings`).
```
```suggestion
Can only be configured for all teams (`org_settings`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+#### integrations
+
+The `integrations` section lets you define calendar events and ticket settings for failing policy and vulnerablity automations. Learn more about automations in Fleet [here](../Using%20Fleet/Automations.md).
```
```suggestion
The `integrations` section lets you define calendar events and ticket settings for failing policy and vulnerablity automations. Learn more about automations in Fleet [here](https://fleetdm.com/docs/using-fleet/automations).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+##### google_calendar
+
+- `api_key_json` is the contents of the JSON file downloaded when you create your Google Workspace service account API key (default: empty).
+- `domain` is the primary domain used to identify your end user's work calendar (default: emtpy).
```
```suggestion
- `api_key_json` is the contents of the JSON file downloaded when you create your Google Workspace service account API key (default: `""`).
- `domain` is the primary domain used to identify your end user's work calendar (default: `""`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+##### jira
+
+- `url` is the URL of your Jira (default: empty)
+- `username` is the username of your Jira account (default: empty).
+- `api_token` is the Jira API token (default: empty).
+- `project_key` is the project key location in your Jira project's URL. For example, in "jira.example.com/projects/EXMPL," "EXMPL" is the project key (default: empty).
```
```suggestion
- `url` is the URL of your Jira (default: `""`)
- `username` is the username of your Jira account (default: `""`).
- `api_token` is the Jira API token (default: `""`).
- `project_key` is the project key location in your Jira project's URL. For example, in "jira.example.com/projects/EXMPL," "EXMPL" is the project key (default: `""`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+##### zendesk
+
+- `url` is the URL of your Zendesk (default: empty)
+- `username` is the username of your Zendesk account (default: empty).
+- `api_token` is the Zendesk API token (default: empty).
+- `group_id`is found by selecting **Admin > People > Groups** in Zendesk. Find your group and select it. The group ID will appear in the search field.
```
```suggestion
  - `url` is the URL of your Zendesk (default: `""`)
- `username` is the username of your Zendesk account (default: `""`).
- `api_token` is the Zendesk API token (default: `""`).
- `group_id`is found by selecting **Admin > People > Groups** in Zendesk. Find your group and select it. The group ID will appear in the search field.
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+#### webhook_settings
+
+The `webhook_settings` section lets you define webhook settings for failing policy, vulnerability, and host status automations. Learn more about automations in Fleet [here](../Using%20Fleet/Automations.md).
```
```suggestion
The `webhook_settings` section lets you define webhook settings for failing policy, vulnerability, and host status automations. Learn more about automations in Fleet [here](https://fleetdm.com/docs/using-fleet/automations).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+##### failing_policies_webhook
+
+- `enable_failing_policies_webhook` (default: `false`)
+- `destination_url` is the URL to `POST` to when the condition for the webhook triggers (default: empty).
```
```suggestion
- `destination_url` is the URL to `POST` to when the condition for the webhook triggers (default: `""`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+##### host_status_webhook
+
+- `enable_host_status_webhook` (default: `false`)
+- `destination_url` is the URL to `POST` to when the condition for the webhook triggers (default: empty).
```
```suggestion
- `destination_url` is the URL to `POST` to when the condition for the webhook triggers (default: `""`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+##### vulnerabilities_webhook
+
+- `enable_vulnerabilities_webhook` (default: `false`)
+- `destination_url` is the URL to `POST` to when the condition for the webhook triggers (default: empty).
```
```suggestion
- `destination_url` is the URL to `POST` to when the condition for the webhook triggers (default: `""`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
````diff
@@ -0,0 +1,548 @@
+      host_batch_size: 0
+```
+
+Can only be configure for all teams (`org_settings`).
````
```suggestion
Can only be configured for all teams (`org_settings`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
```diff
@@ -0,0 +1,548 @@
+
+The `mdm` section lets you enable MDM features in Fleet.
+
+- `apple_bm_default_team` - is name of the team that macOS hosts in Apple Business Manager automatically enroll to when they're first setup. If empty, hosts will enroll to "No team" (default: empty).
```
```suggestion
- `apple_bm_default_team` - is name of the team that macOS hosts in Apple Business Manager automatically enroll to when they're first set up. If empty, hosts will enroll to "No team" (default: `""`).
```

rachaelshaw commented on docs/Using Fleet/GitOps.md:
````diff
@@ -0,0 +1,548 @@
+    apple_bm_default_team: "Workstations" # Available in Fleet Premium
+```
+
+Can only be configure for all teams (`org_settings`).
````
```suggestion
Can only be configured for all teams (`org_settings`).
```

noahtalerman commented:
Thanks for the review @rachaelshaw! I think I addressed all your feedback.

Note to self to run big changes like this through Grammarly (or similar grammar check) to catch the easy ones.
